### DISCUSSION TITLE
py5 exceptions policy - Which Processing exceptions can py5 catch and rephrase?

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#370] [Q&A] [answered]
Cheers @hx2A !

Should I report an issue and propose a message when I get an exception from Processing that maybe could have a more helpful message for learners?

For instance, if you try to use a `bezier_vertex()` outside a `begin_shape()` context:

```
py5 encountered an error in your code:

File "PSurfaceNone.java", line 356, in processing.core.PSurfaceNone$AnimationThread.run

Exception: Java Exception

The above exception was the direct cause of the following exception:
...
File "/home/villares/.local/lib/python3.10/site-packages/villares/arcs.py", line 135, in b_arc
    bezier_vertex(px2, py2, px1, py1, px0, py0)

java.lang.RuntimeException: java.lang.RuntimeException: beginShape() or beginShape(POLYGON) must be used before bezierVertex() or quadraticVertex()
```

Could perhaps become something like (if it is feasible):

``` 
py5 encountered an error in your code:

File "/home/villares/.local/lib/python3.10/site-packages/villares/arcs.py", line 135, in b_arc
    bezier_vertex(px2, py2, px1, py1, px0, py0)

The begin_shape() method must be called before using bezier_vertex() or quadratic_vertex()
```

**Comment by hx2A** [comment]
py5's custom exception handing is [here](https://github.com/py5coding/py5generator/blob/main/py5_resources/py5_module/py5/custom_exceptions.py), currently handling all `TypeError` and `NameError` exceptions. You can try registering new exception messages with the (undocumented) function `py5.register_exception_msg()`.

My vision for this was for it to be expanded to more exceptions, kind of like p5's friendly error system. Also, for it to be easily customizable by educators for the specific needs of their students. For example, you might want to customize it to provide exceptions in portuguese instead of english, or maybe have exceptions that provide URLs to documentation you create for a class.

However, looking at the example you have brought up, I see it doesn't handle what you want so well. It could work, but now you'd need to manage all Java Exceptions messages, not just the ones caused by using methods outside of the begin/end shape block. I see ways that this functionality could be improved.

How about we revisit this after the next release? I could make some improvements so it can achieve its potential, and you use it to customize the exceptions? We can come up with something that handles this and many other exceptions as well.

**Reply by villares** [reply]
Thanks! I'm in no hurry about this.

Another question/idea: where do the Processing maintainers keep those messages? Is it dispersed throughout the Java code base? Could/should we "scrap" those into a CSV for fiddling?

I'm thinking about your automation strategies that help us build upon Processing while not getting mad keeping track of changes up there.

**Reply by villares** [reply]
My naive intuition would be, we could have a database of Java exceptions we want to rephrase, if it's not there, just serve the original Java exception, would that be possible?

**Reply by hx2A** [reply, marked as answer]
> where do the Processing maintainers keep those messages? Is it dispersed throughout the Java code base? Could/should we "scrap" those into a CSV for fiddling?

The error messages are all over the place, but for a lot of them there is a pattern to it. It wouldn't be that hard to extract all of them because we can look for the places where exceptions are raised.

> we could have a database of Java exceptions we want to rephrase, if it's not there, just serve the original Java exception, would that be possible?

Right, that will be a good approach.
